presto-common/src/main/java/com/facebook/presto/common/type/DistinctType.java

```diff
@@ -0,0 +1,340 @@
+        }
+        else {
+            this.superType = Optional.empty();
+        }
+    }
+
+    public UserDefinedDistinctType getUserDefinedDistinctType()
+    {
+        return userDefinedDistinctType;
+    }
+
+    public QualifiedObjectName getName()
+    {
+        return userDefinedDistinctType.getName();
+    }
+
+    public Type getBase()
+    {
+        return baseType;
+    }
+
+    public Optional<DistinctType> getSuperTypeLoadIfNeeded()
+    {
+        if (!getSuperTypeName().isPresent() || superType.isPresent()) {
+            return superType;
+        }
+        superType = Optional.of((DistinctType) typeManager.getType(new TypeSignature(getSuperTypeName().get())));
+        userDefinedDistinctType = new UserDefinedDistinctType(
+                userDefinedDistinctType.getName(),
+                userDefinedDistinctType.getBase(),
+                userDefinedDistinctType.getSuperTypeName(),
+                Optional.of(superType.get().getUserDefinedDistinctType()),
+                userDefinedDistinctType.isOrderable());
+
+        this.typeSignature = new TypeSignature(userDefinedDistinctType);
+        return superType;
+    }
+
+    public Optional<DistinctType> getSuperTypeIfAvailable()
+    {
+        return superType;
+    }
+
+    public Optional<QualifiedObjectName> getSuperTypeName()
+    {
+        return userDefinedDistinctType.getSuperTypeName();
+    }
+
+    // Computes the lowest common ancestor of 2 distinct types in the type tree.
+    // Visits as few types as possible, in order to avoid network calls to fetch the types.
+    public static Optional<DistinctType> getLowestCommonSuperType(DistinctType firstType, DistinctType secondType)
+    {
+        if (firstType.getBase() != secondType.getBase()) {
+            return Optional.empty();
+        }
+
+        HashMap<QualifiedObjectName, DistinctType> seenTypes = new HashMap<>();
+        Optional<DistinctType> firstAncestor = Optional.of(firstType);
+        Optional<DistinctType> secondAncestor = Optional.of(secondType);
+
+        seenTypes.put(firstAncestor.get().getName(), firstAncestor.get());
+        while (firstAncestor.get().getSuperTypeIfAvailable().isPresent()) {
+            firstAncestor = firstAncestor.get().getSuperTypeIfAvailable();
+            seenTypes.put(firstAncestor.get().getName(), firstAncestor.get());
+        }
+
+        if (seenTypes.containsKey(secondAncestor.get().getName())) {
+            return secondAncestor;
+        }
+        seenTypes.put(secondAncestor.get().getName(), secondAncestor.get());
+
+        while (secondAncestor.get().getSuperTypeIfAvailable().isPresent()) {
+            secondAncestor = secondAncestor.get().getSuperTypeIfAvailable();
+            if (seenTypes.containsKey(secondAncestor.get().getName())) {
+                return secondAncestor;
+            }
+            seenTypes.put(secondAncestor.get().getName(), secondAncestor.get());
+        }
+
```

Comment:
I'm not seeing why we need any more logic after this point. If execution gets past the second while loop we've traversed all the supertypes of  `secondType` without encountering a supertype of `firstType`. Why can't we then immediately return `Optional.empty()`?

Reply:
The second part of this logic does exactly the same, but may do some network fetching. We split it into two parts because distinct type instances are lazily populated, and on the worker one of those instances might not have full heirarchy of all parents computed. So we try to minimize the distinct type parent computations. getParentTypeIfAvailable() vs getParentTypeLoadIfNeeded()

Reply:
> We split it into two parts because distinct type instances are lazily populated, and on the worker one of those instances might not have full heirarchy of all parents computed. 

Ah, that's the part I was missing. Thanks!